producer/s3.go

```diff
@@ -0,0 +1,349 @@
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/trivago/gollum/core"
+	"github.com/trivago/gollum/core/log"
+	"github.com/trivago/gollum/shared"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	s3CredentialEnv    = "environment"
+	s3CredentialStatic = "static"
+	s3CredentialShared = "shared"
+	s3CredentialNone   = "none"
+)
+
+// S3 producer plugin
+// This producer sends data to an AWS S3 Bucket.
+// Configuration example
+//
+//  - "producer.S3":
+//    Region: "eu-west-1"
+//    Endpoint: "s3-eu-west-1.amazonaws.com"
+//    StorageClass: "STANDARD"
+//    CredentialType: "none"
+//    CredentialId: ""
+//    CredentialToken: ""
+//    CredentialSecret: ""
+//    CredentialFile: ""
+//    CredentialProfile: ""
+//    BatchMaxMessages: 5000
+//    ObjectMaxMessages: 5000
+//    ObjectMessageDelimiter: "\n"
+//    SendTimeframeMs: 10000
+//    BatchTimeoutSec: 30
+//    TimestampWrite: "2006-01-02T15:04:05"
+//    StreamMapping:
+//      "*" : "default"
+//
+// Region defines the amazon region of your s3 bucket.
+// By default this is set to "eu-west-1".
+//
+// Endpoint defines the amazon endpoint for your s3 bucket.
+// By default this is set to "s3-eu-west-1.amazonaws.com"
+//
+// Endpoint defines the amazon s3 storage class for objects created.
+// By default this is set to "STANDARD"
+//
+// CredentialType defines the credentials that are to be used when
+// connectiong to kensis. This can be one of the following: environment,
+// static, shared, none.
+// Static enables the parameters CredentialId, CredentialToken and
+// CredentialSecret shared enables the parameters CredentialFile and
+// CredentialProfile. None will not use any credentials and environment
+// will pull the credentials from environmental settings.
+// By default this is set to none.
+//
+// BatchMaxMessages defines the maximum number of messages to upload per
+// batch. By default this is set to 5000.
+//
+// ObjectMaxMessages defines the maximum number of messages to join into
+// an s3 object. By default this is set to 5000.
+//
+// ObjectMessageDelimiter defines the string to delimit messages within
+// an s3 object. By default this is set to "\n".
+//
+// SendTimeframeMs defines the timeframe in milliseconds in which a second
+// batch send can be triggered. By default this is set to 10000, i.e. ten
+// upload operations per second per s3 path.
+//
+// BatchTimeoutSec defines the number of seconds after which a batch is
+// flushed automatically. By default this is set to 30.
+//
+// TimestampWrite defines the go timestamp format that will be used in naming
+// objects. Objects are named <s3_path><timestamp><sha1>. By default timestamp
+// is set to "2006-01-02T15:04:05".
+//
+// StreamMapping defines a translation from gollum stream to s3 path. If
+// no mapping is given the gollum stream name is used as s3 path.
```

Comment:
You should note that the path is not "s3://bucket/path" but only "bucket/path" or "bucket".
It does make sense to change the example above to "*" : "bucket/path", too.